CNV-68646: Golden images in heterogenous multi-arch clusters is now GA (#113596)
openshift/openshift-docs: modules/virt-about-architecture-defaults-heterogeneous-clusters.adoc

// Module included in the following assemblies:
//
// * virt/creating_vms_advanced/virt-boot-source-image-heterogeneous-clusters.adoc

:_mod-docs-content-type: CONCEPT
[id="virt-about-architecture-defaults-heterogeneous-clusters_{context}"]
= Default architecture behavior in heterogeneous clusters

[role="_abstract"]

> In a heterogeneous cluster, {VirtProductName} creates architecture-specific data sources for each boot source image so that VMs run on the correct architecture regardless of which node they are scheduled on.

Reviewer: ```suggestion
In a heterogeneous cluster, worker nodes can run different CPU architectures, such as amd64 and arm64. A boot source image built for one architecture cannot run on a node with a different architecture. If a virtual machine uses a boot source that does not match the architecture of the node it is scheduled on, the VM can fail to start or not run as expected.

To prevent this mismatch, {VirtProductName} creates a separate boot source for each supported architecture. Each boot source is represented by a DataSource object, so you can create VMs for a specific architecture regardless of which node they are scheduled on.
```

Reviewer: do we need to add a block that warns from NOT enabling the FG since it can lead to vm boot failures? @MirzWeiss

== How data sources for boot source images work in heterogeneous clusters

> After you enable the `enableMultiArchBootImageImport` feature gate, the Scheduling, Scale, and Performance (SSP) Operator creates a separate `DataSource` object for each supported architecture. For example, the `rhel9` boot source image gets the following data sources:

Reviewer: ```suggestion
After you enable the `enableMultiArchBootImageImport` feature gate, the Scheduling, Scale, and Performance (SSP) Operator creates a new separate boot source for each supported architecture, represented by a DataSource object. For example, the rhel9 boot source image gets
```

* `rhel9-amd64`
* `rhel9-arm64`

> The original name, in this example `rhel9`, points to the data source that matches the control-plane architecture. If the control-plane nodes use the `amd64` architecture, `rhel9` points to `rhel9-amd64`.

Reviewer: ```suggestion
The original name, in this example `rhel9`, resolves to the boot source that matches the control-plane architecture. If control-plane nodes use amd64, `rhel9` references `rhel9-amd64``.
```

Existing VM manifests and templates that use the unsuffixed name continue to work without changes.

== Default architecture for standalone VMs

A VM that is not based on a boot source image, for example a VM from a container disk, HTTP source, upload, or clone, defaults to the control-plane architecture. The `spec.template.spec.architecture` field in the `VirtualMachine` manifest controls which architecture the VM uses.

== Default architecture for standalone data volumes

> A `DataVolume` from a registry source outside the boot source image pipeline does not have an architecture by default. Without an explicit architecture, CDI pulls whichever image variant the registry returns. The `spec.source.registry.platform.architecture` field in the `DataVolume` manifest controls which architecture CDI pulls. For boot source images managed through the `HyperConverged` CR, the `ssp.kubevirt.io/dict.architectures` annotation selects the correct architecture automatically.

Reviewer: pipeline is a bit confusing.
how about:

```suggestion
A `DataVolume` from a registry source outside the boot source images managed through the HyperConverged CR does not have an architecture by default. Without an explicit architecture, CDI pulls whichever image variant the registry returns. The `spec.source.registry.platform.architecture` field in the `DataVolume` manifest controls which architecture CDI pulls. For boot source images managed through the `HyperConverged` CR, the `ssp.kubevirt.io/dict.architectures` annotation selects the correct architecture automatically.
```

====

== How data sources for boot source images work in heterogeneous clusters

After you enable the `enableMultiArchBootImageImport` feature gate, the Scheduling, Scale, and Performance (SSP) Operator creates a new separate boot source for each supported architecture, represented by a `DataSource` object. For example, the `rhel9` boot source image gets the following data sources:

* `rhel9-amd64`
* `rhel9-arm64`

The original name, such as `rhel9`, defaults to the boot source that matches the control-plane architecture. For example, if the control-plane nodes use `amd64`, then `rhel9` resolves to `rhel9-amd64`.

Existing VM manifests and templates that use the unsuffixed name continue to work without changes.

== Default architecture for standalone VMs

A VM that is not based on a boot source image, for example a VM from a container disk, HTTP source, upload, or clone, defaults to the control-plane architecture. The `spec.template.spec.architecture` field in the `VirtualMachine` manifest controls which architecture the VM uses.

== Default architecture for standalone data volumes

> A `DataVolume` created directly from a registry source, rather than through the `HyperConverged` CR, does not have an architecture by default. Without an explicit architecture, CDI pulls whichever image variant the registry returns. The `spec.source.registry.platform.architecture` field in the `DataVolume` manifest controls which architecture CDI pulls. For boot source images managed through the `HyperConverged` CR, the `ssp.kubevirt.io/dict.architectures` annotation selects the correct architecture automatically.

Reviewer 2: ```suggestion
A `DataVolume` created directly from a registry source does not have a default architecture. Without an explicit architecture, CDI pulls whichever image variant the registry returns. Populate the `spec.source.registry.platform.architecture` field in the `DataVolume` manifest to control which architecture to pull.
```
